app/controllers/services_controller.rb

```diff
@@ -1,16 +1,18 @@
 class ServicesController < ApplicationController
-  def register
-    service = Service.find_or_create_by url: params.require(:url)
-    params.require(:models).each do |model_data|
-      model = Model.find_by name: model_data.require(:name)
-      if model.present?
-        head :unprocessable_entity and return if model.service != service
-      else
-        Model.create(name: model_data.require(:name), service: service)
+  include ServiceChangeNotifier
+    def register
+      service = Service.find_or_create_by url: params.require(:url)
+      params.require(:models).each do |model_data|
+        model = Model.find_by name: model_data.require(:name)
+        if model.present?
+          head :unprocessable_entity and return if model.service != service
+        else
+          Model.create(name: model_data.require(:name), service: service)
+        end
       end
-    end
-    render json: service,
-           only: :url,
-           include: { models: { only: :name } }
+      render json: service,
+             only: :url,
+             include: { models: { only: :name } }
+      ServiceChangeNotifier.notify_services_of_changes if (service.changed? || service.models.each{|m| m.changed?}.present?)
```

Comment:
Your intuition is correct: take a look at the [docs](http://apidock.com/rails/ActiveModel/Dirty/changed%3F). `changed?` looks at unsaved changes — because that's really all it can do. You can imagine how it might be difficult to programmatically express the "just" in "did we just make a change to an object" in a way that  works consistently.

As far as `map` and `each`, that was going to be an interview question :-P

@enragednuke explained it very well — I would just add that it's the difference between procedural and functional statements. `each` is for doing things without caring about its return value; `map` is for calling methods that don't do anything but whose return value is significant.